File: java/src/DateUtils.java
// DateUtils.java
// Various utilities to work with dates
import java.util.ArrayList;
import java.util.Calendar;
import java.util.Date;
import java.util.TimeZone;
import java.util.concurrent.TimeUnit;

public final class DateUtils {
	private DateUtils(){}
	
	private static Calendar _getCalendar()
	{
        TimeZone tz	= TimeZone.getTimeZone("UTC");
        return Calendar.getInstance(tz);
	}
	
	private static Date _addPeriod(Date startDate, int periodType, int numberOfPeriods)
	{
		Calendar cldr = _getCalendar();
        cldr.setTime(startDate);
        cldr.add(periodType, numberOfPeriods);
        return cldr.getTime();
	}
	
	public static Date AddDays(Date startDate, int numDays)
	{
        return _addPeriod(startDate, Calendar.DATE, numDays);
	}
	
	public static Date AddMonths(Date startDate, int numMonths)
	{
        return _addPeriod(startDate, Calendar.MONTH, numMonths);
	}
	
	public static Date AddYears(Date startDate, int numYears)
	{
        return _addPeriod(startDate, Calendar.YEAR, numYears);
	}
	
	public static long CountDays(Date startDate, Date endDate)
	{
    	long milliSecs = endDate.getTime() - startDate.getTime();
    	long nbDays = TimeUnit.DAYS.convert(milliSecs, TimeUnit.MILLISECONDS);
    	return nbDays;
	}
	
    public static int CountYears(Date startDate, Date endDate)
    {
        int years = 0;
        startDate = AddYears(startDate, 1);
        while (!startDate.after(endDate))
        {
            years++;
            startDate = AddYears(startDate, 1);
        }
        return years;
    }
    
    public static int GetDay(Date d)
    {
    	return _getDateComponent(d, Calendar.DAY_OF_MONTH);
    }
    
    public static int GetMonth(Date d)
    {
    	return _getDateComponent(d, Calendar.MONTH);
    }
    
    private static int _getDateComponent(Date d, int comp)
    {
    	Calendar cldr = _getCalendar();
    	cldr.setTime(d);
    	return cldr.get(comp);
    }
    
	public static Date AddPeriod(Date startDate, String period)
    {
        // Parse period
        int num = Integer.parseInt(period.substring(0, period.length() - 1));
        String unit = period.substring(period.length() - 1);

        // Start date adjustment
        Calendar cldr = _getCalendar();
        cldr.setTime(startDate);
        int dayOfWeek = cldr.get(Calendar.DAY_OF_WEEK);
        
        if (dayOfWeek == 0 || dayOfWeek == 6)
        {
            /* Sunday = 0, Saturday = 6
             * If Sunday (0), add 1 day
             * If Saturday (6), add 2 days
             * day of week / 6 + 1 is 1 for Sunday, 2 for Saturday */
            startDate = AddDays(startDate, dayOfWeek / 6 + 1);
        }

        Date endDate = startDate;
        switch (unit)
        {
            case "D":
                endDate = AddDays(startDate, num);
                break;
            case "W":
                endDate = AddDays(startDate, num * 7);
                break;
            case "M":
                endDate = AddMonths(startDate, num);
                break;
            default:    // Assume year
                endDate = AddYears(startDate, num);
                break;
        }

        // End date adjustment
        cldr.setTime(endDate);
        dayOfWeek = cldr.get(Calendar.DAY_OF_WEEK);
        if (dayOfWeek == 0 || dayOfWeek == 6)
        {
        	endDate = AddDays(endDate, dayOfWeek / 6 + 1);
        }

        return endDate;
    }
	
	public static double YearFrac(Date startDate, Date endDate)
	{
		return YearFrac(startDate, endDate, DaycountConvention.ActualActual);
	}
	
    public static double YearFrac(Date startDate, Date endDate, DaycountConvention dayCount)
    {
        double yearBasis = dayCount == DaycountConvention.Actual365 ? 365.0 : 360.0;
        if (dayCount == DaycountConvention.ActualActual)
        {
            yearBasis = CountDays(startDate, AddYears(startDate, 1));
        }

        // Find number of whole years
        int wholeYears = CountYears(endDate, startDate);

        if(wholeYears > 0)
        {
            startDate = AddYears(startDate, wholeYears);
        }

        // Loop through months
        int totalNumberOfDays = 0;
        while (startDate.before(endDate))
        {
            Date monthStartDate = AddDays(startDate, -(GetDay(startDate)-1));
            Date monthEndDate = endDate;
            if (startDate.getTime() != monthStartDate.getTime())    // if start date is not first day of month
            {
                monthEndDate = AddMonths(monthStartDate, 1);
            }
            else if (GetMonth(startDate) != GetMonth(endDate))  // If end date is not in same month, then find last day
            {
            	monthEndDate = (AddMonths(monthStartDate, 1).after(endDate)) ? endDate : AddMonths(monthStartDate, 1);
            }

            // Increment total by number of days in (partial) month following convention
            if (dayCount == DaycountConvention.Thirty360)
            {
                int numberOfDays = 31 - GetDay(startDate);
                if (GetMonth(startDate) == 2 /*month is February*/ && numberOfDays >= 28)
                {
                    totalNumberOfDays += 30;
                }
                else
                {
                    totalNumberOfDays += Math.min(numberOfDays, 30);
                }
            }
            else
            {
                totalNumberOfDays += CountDays(startDate, monthEndDate);
            }

            // Increment start date and repeat
            if (startDate != monthStartDate)
            {
                startDate = monthEndDate;
            }
            else
            {
                startDate = AddMonths(startDate, 1);
            }

        }

        return (double)wholeYears + (totalNumberOfDays / yearBasis);
    }
    
    public static Period CalculateZcDates(Date valueDate, String period)
    {
        Date startDate = valueDate;
        Period zcDates = new Period();
        switch (period)
        {
            case "0D":
            	zcDates.setStartDate(valueDate);
            	zcDates.setEndDate(valueDate);
                break;
            case "ON":
            	zcDates.setStartDate(startDate);
            	zcDates.setEndDate(AddPeriod(startDate, "1D"));
                break;
            case "TN":
                startDate = AddPeriod(startDate, "1D");
            	zcDates.setStartDate(startDate);
            	zcDates.setEndDate(AddPeriod(startDate, "1D"));
                break;
            case "SW":
                startDate = AddPeriod(startDate, "2D");
            	zcDates.setStartDate(startDate);
            	zcDates.setEndDate(AddPeriod(startDate, "1W"));
                break;
            default:    // Spot. Assume 2 days spot lag. TODO: for LIBOR use 0 days when underlying is swap or LIBOR
                startDate = AddPeriod(startDate, "2D");
            	zcDates.setStartDate(startDate);
            	zcDates.setEndDate(AddPeriod(startDate, period));
                break;
        }
        return zcDates;
    }
    
    public static ArrayList<Coupon> GetCouponDates(Date settlementDate, Date maturityDate, Frequency frequency)
    {
    	ArrayList<Coupon> coupons = new ArrayList<Coupon>();
    	Period couponDates = new Period(AddMonths(maturityDate, -frequency.getValue()), maturityDate);
        while (!(couponDates.getStartDate().before(settlementDate)))
        {
        	coupons.add(new Coupon(couponDates.getStartDate(), couponDates.getEndDate(), DaycountConvention.Actual360));
            couponDates = new Period(AddMonths(couponDates.getStartDate(), -frequency.getValue()), couponDates.getStartDate());
        }
        return coupons;
    }}

File: csharp/vanilla_swap/vanilla_swap/DateUtils.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace VanillaSwap
{
    public static class DateUtils
    {
        /// <summary>
        /// Calculates the number of years as a fraction, using period-based calculation and the given daycount convention.
        /// </summary>
        /// <param name="startDate">Start date of the period</param>
        /// <param name="endDate">End date of the period</param>
        /// <param name="dayCount">Daycount basis, e.g. Actual/Actual or 30/360</param>
        /// <returns>The number of years between start and end date</returns>
        public static double YearFrac(DateTime startDate, DateTime endDate, DayCountBasis dayCount = DayCountBasis.ActualActual)
        {
            var yearBasis = dayCount == DayCountBasis.Actual365 ? 365.0 : 360.0;
            if (dayCount == DayCountBasis.ActualActual)
            {
                yearBasis = (startDate.AddYears(1) - startDate).Days;
            }

            // Find number of whole years
            var wholeYears = CountYears(endDate, startDate);

            if(wholeYears > 0)
            {
                startDate = startDate.AddYears(wholeYears);
            }

            // Loop through months
            var totalNumberOfDays = 0;
            while (startDate <= endDate)
            {
                var monthStartDate = startDate.AddDays(-(startDate.Day - 1));
                var monthEndDate = endDate;
                if (startDate != monthStartDate)    // if start date is not first day of month
                {
                    monthEndDate = monthStartDate.AddMonths(1);
                }
                else if (startDate.Month != endDate.Month)  // If end date is not in same month, then find last day
                {
                    monthEndDate = new DateTime(Math.Min(endDate.Ticks, startDate.AddMonths(1).Ticks));
                }

                // Increment total by number of days in (partial) month following convention
                if (dayCount == DayCountBasis.Thirty360)
                {
                    var numberOfDays = 31 - startDate.Day;
                    if (startDate.Month == 2 /*month is February*/ && numberOfDays >= 28)
                    {
                        totalNumberOfDays += 30;
                    }
                    else
                    {
                        totalNumberOfDays += Math.Min(numberOfDays, 30);
                    }
                }
                else
                {
                    totalNumberOfDays += (monthEndDate - startDate).Days;
                }

                // Increment start date and repeat
                if (startDate != monthStartDate)
                {
                    startDate = monthEndDate;
                }
                else
                {
                    startDate = startDate.AddMonths(1);
                }

            }

            return (double)wholeYears + (totalNumberOfDays / yearBasis);
        }

        public static int CountYears(DateTime startDate, DateTime endDate)
        {
            int years = 0;
            startDate = startDate.AddYears(1);
            while (startDate <= endDate)
            {
                years++;
                startDate = startDate.AddYears(1);
            }
            return years;
        }

        public static IEnumerable<Coupon> GetCouponDates(DateTime settlementDate, DateTime maturityDate, Frequency frequency)
        {
            var couponDates = new Tuple<DateTime, DateTime>(maturityDate.AddMonths(-(int)frequency), maturityDate);
            while (couponDates.Item1 > settlementDate)
            {
                yield return new Coupon(couponDates.Item1, couponDates.Item2, DayCountBasis.Actual360);
                couponDates = new Tuple<DateTime, DateTime>(couponDates.Item1.AddMonths(-(int)frequency), couponDates.Item1);
            }
        }

        /// <summary>
        /// Basic add period which does not take into account date adjustment conventions such as modified following.
        /// </summary>
        /// <param name="startDate"></param>
        /// <param name="period"></param>
        /// <returns></returns>
        public static DateTime AddPeriod(DateTime startDate, string period)
        {
            // Parse period
            var num = int.Parse(period.Substring(0, period.Length - 1));
            var unit = period[period.Length - 1].ToString();

            // Start date adjustment
            if (startDate.DayOfWeek == DayOfWeek.Saturday || startDate.DayOfWeek == DayOfWeek.Sunday)
            {
                /* Sunday = 0, Saturday = 6
                 * If Sunday (0), add 1 day
                 * If Saturday (6), add 2 days
                 * day of week / 6 + 1 is 1 for Sunday, 2 for Saturday */
                startDate.AddDays((int)startDate.DayOfWeek / 6 + 1);
            }

            var endDate = startDate;
            switch (unit)
            {
                case "D":
                    endDate = startDate.AddDays(num);
                    break;
                case "W":
                    endDate = startDate.AddDays(num * 7);
                    break;
                case "M":
                    endDate = startDate.AddMonths(num);
                    break;
                default:    // Assume year
                    endDate = startDate.AddYears(num);
                    break;
            }

            // End date adjustment
            if (endDate.DayOfWeek == DayOfWeek.Saturday || endDate.DayOfWeek == DayOfWeek.Sunday)
            {
                endDate.AddDays((int)endDate.DayOfWeek / 6 + 1);
            }

            return endDate;
        }

        public static Tuple<DateTime, DateTime> CalculateZcDates(DateTime valueDate, string period)
        {
            var startDate = valueDate;
            var endDate = valueDate;

            switch (period)
            {
                case "0D":
                    return new Tuple<DateTime, DateTime>(startDate,endDate);
                case "ON":
                    return new Tuple<DateTime, DateTime>(startDate, AddPeriod(startDate, "1D"));
                case "TN":
                    startDate = AddPeriod(startDate, "1D");
                    return new Tuple<DateTime, DateTime>(startDate, AddPeriod(startDate, "1D"));
                case "SW":
                    startDate = AddPeriod(startDate, "2D");
                    return new Tuple<DateTime, DateTime>(startDate, AddPeriod(startDate, "1W"));
                default:    // Spot. Assume 2 days spot lag. TODO: for LIBOR use 0 days when underlying is swap or LIBOR
                    startDate = AddPeriod(startDate, "2D");
                    return new Tuple<DateTime, DateTime>(startDate, AddPeriod(startDate, period));
            }
        }
    }
}
